power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -0,0 +1,471 @@
+// SPDX-FileCopyrightText: 2022 Contributors to the Power Grid Model project <dynamic.grid.calculation@alliander.com>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_AUXILIARY_SERIALIZTION_DESERIALIZER_HPP
+#define POWER_GRID_MODEL_AUXILIARY_SERIALIZTION_DESERIALIZER_HPP
+
+#include "../../exception.hpp"
+#include "../../power_grid_model.hpp"
+#include "../meta_data.hpp"
+#include "../meta_data_gen.hpp"
+
+#include <nlohmann/json.hpp>
+
+#include <msgpack.hpp>
+
+#include <set>
+#include <span>
+#include <sstream>
+#include <string_view>
+
+// converter for double[3]
+namespace msgpack {
+MSGPACK_API_VERSION_NAMESPACE(MSGPACK_DEFAULT_API_NS) {
+    namespace adaptor {
+    template <> struct convert<power_grid_model::RealValue<false>> {
+        msgpack::object const& operator()(msgpack::object const& o, power_grid_model::RealValue<false>& v) const {
+            if (o.type != msgpack::type::ARRAY) {
+                throw msgpack::type_error();
+            }
+            if (o.via.array.size != 3) {
+                throw msgpack::type_error();
+            }
+            for (int8_t i = 0; i != 3; ++i) {
+                if (o.via.array.ptr[i].is_nil()) {
+                    continue;
+                }
+                o.via.array.ptr[i] >> v(i);
+            }
+            return o;
+        }
+    };
+
+    template <class T>
+        requires(std::same_as<T, msgpack::object> || std::same_as<T, msgpack::object_kv>)
+    struct convert<std::span<const T>> {
+        msgpack::object const& operator()(msgpack::object const& o, std::span<const T>& span) const {
+            if constexpr (std::same_as<T, msgpack::object>) {
+                span = {o.via.array.ptr, o.via.array.size};
+            } else {
+                span = {o.via.map.ptr, o.via.map.size};
+            }
+            return o;
+        }
+    };
+
+    } // namespace adaptor
+} // MSGPACK_API_VERSION_NAMESPACE(MSGPACK_DEFAULT_API_NS)
+} // namespace msgpack
+
+namespace power_grid_model::meta_data {
+
+constexpr struct from_msgpack_t {
+} from_msgpack;
+constexpr struct from_json_t {
+} from_json;
```

Comment:
Maybe more readable:

* split it up into two
* don't specify `from_` so that it can be re-used between the serializer and deserializer, so that we can use the same tag to do a full cycle test. in stead, specify that it's a `_tag`

```cpp
struct json_tag_t{};
struct msgpack_tag_t{};

constexpr auto json_tag = json_tag_t{};
constexpr auto msgpack_tag = msgpack_tag_t{};
```

in addition: i'm surprised that the formatter doesn't make a single line statement of this

Reply:
1. A tag type does not need to end with `_tag`. `std::piecewise_construct` does not have `_tag` at the end.
2. I think `from_json` is a good name. You don't need this for serializer as you can have actual named functions. For deserializer you need the tag type in the constructor.
3. I will define the struct and value separately.